presto-main/src/main/java/com/facebook/presto/server/DynamicFilterService.java

```diff
@@ -17,28 +17,39 @@
 
 import javax.annotation.concurrent.Immutable;
 import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
 
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
 import static com.google.common.base.Preconditions.checkState;
 import static java.util.Objects.requireNonNull;
 
+@ThreadSafe
 public class DynamicFilterService
 {
     private final Map<SourceDescriptor, DynamicFilterSummaryWithSenders> dynamicFilterSummaries = new HashMap<>();
+    private final Map<SourceDescriptor, SummaryCompleteListener> listeners = new ConcurrentHashMap<>();
```

Comment:
Let me ask the question other way around: why future instead of listener? ;)
Don't feel strong about it.